File: clients/client-js/client-v2/streams/helpers/rateLimiter.js
/**
 * Rate-limiter helpers — composable token-bucket / sliding-window gate
 * factories that return a function suitable for `.gate(fn)` on a Stream.
 *
 * The point of these helpers is purely ergonomic: the user shouldn't have to
 * re-write the refill math every time. The semantics (per-key state in
 * queen_streams.state, partial-ack on deny, FIFO order on lease expiry) all
 * come from the underlying Stream `.gate()` runtime — these factories only
 * decide HOW each request consumes from the bucket.
 *
 * All four common rate-limit shapes are expressible by varying `costFn`:
 *
 *   A) "100 req/sec, batches arbitrary":    costFn = () => 1            on a queue of REQUESTS
 *   B) "100 msg/sec, in any number of req": costFn = () => 1            on a queue of MESSAGES
 *   C) "100 req/sec, exactly 1 msg/req":    costFn = () => 1            same as B (degenerate)
 *   D) "100 weight/sec, cost varies":       costFn = req => req.weight  on a queue of REQUESTS
 *
 * Two gates can be chained on the same pipeline (req/s + msg/s simultaneously
 * required by some OTAs) by running them as two consecutive Stream queries
 * with an intermediate sink queue.
 *
 * Sizing rule (also enforced by tokenBucketGate at start time, and warned
 * by the assertion below): the maximum sustainable rate the bucket can
 * deliver is `min(refillPerSec, capacity / leaseSec)`. Pick capacity and
 * leaseSec so the second term doesn't artificially cap you below
 * refillPerSec — usually `capacity ≈ refillPerSec × leaseSec`.
 */

/**
 * Token-bucket gate factory.
 *
 * Returns a `(msg, ctx) => boolean` function suitable for `.gate(fn)`.
 * The bucket state lives in `ctx.state` (which the runtime persists per-key
 * in queen_streams.state on every ALLOWED message).
 *
 * @param {object} opts
 * @param {number}                 opts.capacity        Max tokens in the bucket (= max burst).
 * @param {number}                 opts.refillPerSec    Steady-state refill rate (tokens / sec).
 * @param {(msg:any)=>number}     [opts.costFn]         Maps a message to its cost in tokens.
 *                                                       Defaults to `() => 1` (one token per message).
 * @param {boolean}               [opts.allowZeroCost]  If true, costFn() may return 0 to mean
 *                                                       "always allow, no token consumed". Defaults true.
 * @returns {(msg:any, ctx:object)=>boolean}
 */
export function tokenBucketGate({
  capacity,
  refillPerSec,
  costFn = () => 1,
  allowZeroCost = true
} = {}) {
  if (!Number.isFinite(capacity) || capacity <= 0) {
    throw new Error('tokenBucketGate: capacity must be a positive number')
  }
  if (!Number.isFinite(refillPerSec) || refillPerSec <= 0) {
    throw new Error('tokenBucketGate: refillPerSec must be a positive number')
  }
  if (typeof costFn !== 'function') {
    throw new Error('tokenBucketGate: costFn must be a function')
  }
  return function tokenBucket(msg, ctx) {
    const now = ctx.streamTimeMs
    if (typeof ctx.state.tokens !== 'number') ctx.state.tokens = capacity
    if (typeof ctx.state.lastRefillAt !== 'number') ctx.state.lastRefillAt = now
    // Refill based on elapsed wall-clock time. Capped at capacity so we
    // never exceed the configured burst.
    const elapsedSec = Math.max(0, (now - ctx.state.lastRefillAt) / 1000)
    ctx.state.tokens = Math.min(capacity, ctx.state.tokens + elapsedSec * refillPerSec)
    ctx.state.lastRefillAt = now

    let cost = costFn(msg)
    if (!Number.isFinite(cost) || cost < 0) cost = 1
    if (cost === 0) return allowZeroCost   // never consume, never block

    if (ctx.state.tokens >= cost) {
      ctx.state.tokens -= cost
      // Lightweight observability: the SDK persists ctx.state on ALLOW only,
      // so these counters automatically reflect "what the bucket actually let
      // through" and can be inspected via SQL on queen_streams.state.
      ctx.state.allowedTotal = (ctx.state.allowedTotal || 0) + 1
      ctx.state.consumedTotal = (ctx.state.consumedTotal || 0) + cost
      return true
    }
    // On deny, ctx.state mutations are discarded by the runtime — so
    // deniedSeen is "lost" and would only be useful if the runtime
    // persisted on every eval. We leave it OUT here so callers don't
    // misinterpret the persisted counters.
    return false
  }
}

/**
 * Sliding-window approximation gate (for "max N events in last W seconds"
 * style limits — e.g. SendGrid daily quotas, OTA hourly quotas).
 *
 * IMPORTANT: this is NOT a precise sliding window. It's a 2-bucket
 * approximation (current + previous) that's accurate within ±1× rate at
 * the boundary. For exact sliding window, you'd need to keep per-event
 * timestamps in state — usually overkill for rate-limiting purposes.
 *
 * @param {object} opts
 * @param {number}                opts.limit          Max events per window.
 * @param {number}                opts.windowSec      Window length in seconds.
 * @param {(msg:any)=>number}    [opts.costFn]        Cost per event (default 1).
 */
export function slidingWindowGate({ limit, windowSec, costFn = () => 1 } = {}) {
  if (!Number.isFinite(limit) || limit <= 0) {
    throw new Error('slidingWindowGate: limit must be a positive number')
  }
  if (!Number.isFinite(windowSec) || windowSec <= 0) {
    throw new Error('slidingWindowGate: windowSec must be a positive number')
  }
  const windowMs = windowSec * 1000
  return function slidingWindow(msg, ctx) {
    const now = ctx.streamTimeMs
    const currentWindow = Math.floor(now / windowMs)
    const elapsedInWindow = (now % windowMs) / windowMs   // 0..1

    if (ctx.state.window !== currentWindow) {
      // Roll: previous window becomes "previous", current resets.
      ctx.state.previousCount = ctx.state.window === currentWindow - 1
        ? (ctx.state.currentCount || 0)
        : 0
      ctx.state.currentCount = 0
      ctx.state.window = currentWindow
    }
    // 2-bucket sliding estimate.
    const estimated = (ctx.state.previousCount || 0) * (1 - elapsedInWindow)
                    + (ctx.state.currentCount  || 0)
    let cost = costFn(msg)
    if (!Number.isFinite(cost) || cost < 0) cost = 1

    if (estimated + cost <= limit) {
      ctx.state.currentCount = (ctx.state.currentCount || 0) + cost
      return true
    }
    return false
  }
}

File: clients/client-py/queen/streams/helpers/rate_limiter.py
"""
Rate-limiter helpers — composable token-bucket / sliding-window gate
factories that return a callable suitable for ``Stream.gate(fn)``.

Mirror of the JS rateLimiter.js helpers. The semantics (per-key state in
``queen_streams.state``, partial-ack on deny, FIFO order on lease expiry)
all come from the underlying Stream gate runtime — these factories only
decide HOW each request consumes from the bucket.

All four common rate-limit shapes are expressible by varying ``cost_fn``:

  A) "100 req/sec, batches arbitrary":     cost_fn = lambda m: 1     on a queue of REQUESTS
  B) "100 msg/sec, in any number of req":  cost_fn = lambda m: 1     on a queue of MESSAGES
  C) "100 req/sec, exactly 1 msg/req":     cost_fn = lambda m: 1     same as B (degenerate)
  D) "100 weight/sec, cost varies":        cost_fn = lambda m: m["weight"]

Sizing rule: max sustainable rate per partition =
    min(refill_per_sec, capacity / lease_sec)
Pick capacity and lease_sec so the second term doesn't artificially cap you
below refill_per_sec — usually capacity ≈ refill_per_sec × lease_sec.
"""

from __future__ import annotations

import math
from typing import Any, Callable


def _is_finite_positive(x: Any) -> bool:
    return (
        isinstance(x, (int, float))
        and not isinstance(x, bool)
        and math.isfinite(x)
        and x > 0
    )


def token_bucket_gate(
    capacity: float,
    refill_per_sec: float,
    cost_fn: Callable[[Any], float] = lambda _msg: 1,
    allow_zero_cost: bool = True,
) -> Callable[[Any, Any], bool]:
    """
    Token-bucket gate factory.

    Returns a ``(msg, ctx) -> bool`` callable suitable for ``Stream.gate(fn)``.
    The bucket state lives in ``ctx.state`` (which the runtime persists per-key
    in ``queen_streams.state`` on every ALLOWED message).
    """
    if not _is_finite_positive(capacity):
        raise ValueError("token_bucket_gate: capacity must be a positive number")
    if not _is_finite_positive(refill_per_sec):
        raise ValueError("token_bucket_gate: refill_per_sec must be a positive number")
    if not callable(cost_fn):
        raise TypeError("token_bucket_gate: cost_fn must be callable")

    def token_bucket(msg: Any, ctx: Any) -> bool:
        now = ctx.stream_time_ms
        state = ctx.state
        if not isinstance(state.get("tokens"), (int, float)):
            state["tokens"] = capacity
        if not isinstance(state.get("lastRefillAt"), (int, float)):
            state["lastRefillAt"] = now

        elapsed_sec = max(0.0, (now - state["lastRefillAt"]) / 1000.0)
        state["tokens"] = min(capacity, state["tokens"] + elapsed_sec * refill_per_sec)
        state["lastRefillAt"] = now

        cost = cost_fn(msg)
        if not isinstance(cost, (int, float)) or isinstance(cost, bool) or not math.isfinite(cost) or cost < 0:
            cost = 1
        if cost == 0:
            return allow_zero_cost

        if state["tokens"] >= cost:
            state["tokens"] -= cost
            state["allowedTotal"] = state.get("allowedTotal", 0) + 1
            state["consumedTotal"] = state.get("consumedTotal", 0) + cost
            return True
        return False

    return token_bucket


def sliding_window_gate(
    limit: float,
    window_sec: float,
    cost_fn: Callable[[Any], float] = lambda _msg: 1,
) -> Callable[[Any, Any], bool]:
    """
    Sliding-window approximation gate (for "max N events in last W seconds").

    NOT a precise sliding window. 2-bucket approximation (current +
    previous) accurate within ±1× rate at the boundary. For exact sliding
    window keep per-event timestamps in state.
    """
    if not _is_finite_positive(limit):
        raise ValueError("sliding_window_gate: limit must be a positive number")
    if not _is_finite_positive(window_sec):
        raise ValueError("sliding_window_gate: window_sec must be a positive number")

    window_ms = window_sec * 1000

    def sliding_window(msg: Any, ctx: Any) -> bool:
        now = ctx.stream_time_ms
        state = ctx.state
        current_window = math.floor(now / window_ms)
        elapsed_in_window = (now % window_ms) / window_ms

        if state.get("window") != current_window:
            state["previousCount"] = (
                state.get("currentCount", 0) if state.get("window") == current_window - 1 else 0
            )
            state["currentCount"] = 0
            state["window"] = current_window

        estimated = (state.get("previousCount", 0) or 0) * (1 - elapsed_in_window) + (
            state.get("currentCount", 0) or 0
        )
        cost = cost_fn(msg)
        if not isinstance(cost, (int, float)) or isinstance(cost, bool) or not math.isfinite(cost) or cost < 0:
            cost = 1

        if estimated + cost <= limit:
            state["currentCount"] = state.get("currentCount", 0) + cost
            return True
        return False

    return sliding_window
